### PR TITLE
Swift 2.3 Migration

### PR DESCRIPTION
**Pull request by jleandroperez**
#### Details:
**Note: This is @sendhil 's work**. I'm just merging in the latest, and sending a PR right right away, so that everyone can take the app for a test ride.

- Install Xcode 8
- Verify that the Unit Tests run correctly
- Take the app for a ride?

Invoking the whole team on this one!. We'll merge on.. two squirrels? three?

Needs review: @astralbodies @aerych @frosty @kurzee @nheagy @diegoreymendez @sendhil  (And me!)

### CODE REVIEW

File: WordPress/Classes/ViewRelated/NUX/SigninHelpers.swift [hunk 1]
```diff
@@ -335,15 +335,15 @@ import Mixpanel
...
 
         if path.isWordPressComPath() {
-            if siteURL.scheme.characters.count == 0 {
+            if siteURL.scheme == nil || siteURL.scheme?.characters.count == 0 {
                 path = "https://\(path)"
             } else if path.rangeOfString("http://") != nil {
                 path = path.stringByReplacingOccurrencesOfString("http://", withString: "https://")
             }
-        } else if siteURL.scheme.characters.count == 0 {
+        } else if siteURL.scheme == nil || siteURL.scheme == nil {
```
Comment (aerych) [inline WordPress/Classes/ViewRelated/NUX/SigninHelpers.swift]: wat?

Comment (aerych) [inline WordPress/Classes/ViewRelated/NUX/SigninHelpers.swift]: Maybe instead something like 
`
        var path = siteURL.absoluteString?.lowercaseString ?? ""
        let scheme = siteURL.scheme ?? ""

        if path.isWordPressComPath() {
            if scheme.characters.count == 0 {
                path = "https://\(path)"
            } else if path.rangeOfString("http://") != nil {
                path = path.stringByReplacingOccurrencesOfString("http://", withString: "https://")
            }
        } else if scheme.characters.count == 0 {
            path = "http://\(path)"
        }

`

File: WordPress/Classes/ViewRelated/Reader/ReaderDetailViewController.swift [hunk 2]
```diff
@@ -449,8 +449,8 @@ public class ReaderDetailViewController : UIViewController, UIViewControllerRest
         var requestURL = url
 
         let absoluteString = requestURL.absoluteString
-        if !absoluteString.hasPrefix("https") {
-            let sslURL = absoluteString.stringByReplacingOccurrencesOfString("http", withString: "https")
+        if !absoluteString!.hasPrefix("https") {
+            let sslURL = absoluteString!.stringByReplacingOccurrencesOfString("http", withString: "https")
```
Comment (aerych) [inline WordPress/Classes/ViewRelated/Reader/ReaderDetailViewController.swift]: Rather than unwrapping maybe change the let statement like this:
``` let absoluteString = requestURL.absoluteString ?? "" ```

File: WordPress/Classes/ViewRelated/Reader/ReaderPostCardCell.swift [hunk 3]
```diff
@@ -351,8 +351,8 @@ import WordPressShared
         var requestURL = url
 
         let absoluteString = requestURL.absoluteString
-        if !(absoluteString.hasPrefix("https")) {
-            let sslURL = absoluteString.stringByReplacingOccurrencesOfString("http", withString: "https")
+        if !(absoluteString!.hasPrefix("https")) {
+            let sslURL = absoluteString!.stringByReplacingOccurrencesOfString("http", withString: "https")
```
Comment (aerych) [inline WordPress/Classes/ViewRelated/Reader/ReaderPostCardCell.swift]: Maybe ``` let absoluteString = requestURL.absoluteString! ``` instead.

**Comment by kwonye** [comment]
TodayWidget isn't even showing up. Investigating...

**Comment by diegoreymendez** [comment]
@jleandroperez - I added some changes to the code to bring in the latests changes from Aztec.

**Comment by kwonye** [comment]
Ok... the Today Widget is now working for me with nothing changed in my project.

@jleandroperez and @diegoreymendez can you verify that running the project adds the WordPress widget? Please try it from a clean install.

**Comment by diegoreymendez** [comment]
@kwonye - It's working perfectly for me.

**Comment by diegoreymendez** [comment]
I updated the development provisioning profile for the Today Widget.

File: Podfile [hunk 4]
```diff
@@ -12,7 +12,7 @@ platform :ios, '9.0'
 abstract_target 'WordPress_Base' do
   pod 'WordPress-iOS-Shared', '0.6.1'
   ## This pod is only being included to support the share extension ATM - https://github.com/wordpress-mobile/WordPress-iOS/issues/5081
-  pod 'WordPressComKit', :git => 'https://github.com/Automattic/WordPressComKit.git', :tag => '0.0.4'
+  pod 'WordPressComKit', :git => 'https://github.com/Automattic/WordPressComKit.git', :commit => "392d7f3326e613b79502ea556f645be161c31a97"
```
Comment (koke) [inline Podfile]: Can we tag a new version (or use a branch) instead of pinning to a commit?

File: Podfile [hunk 5]
```diff
@@ -92,6 +92,17 @@ post_install do |installer_representation|
...
+  #
+  #
+  #
+  #
+
+  installer_representation.pods_project.targets.each do |target|
+    target.build_configurations.each do |config|
+      config.build_settings['SWIFT_VERSION'] = '2.3'
+    end
+  end
```
Comment (koke) [inline Podfile]: This is now between a comment and the thing the comment describes

File: WordPress/Classes/ViewRelated/Media/GravatarOverlayView.swift [hunk 6]
```diff
@@ -27,23 +27,23 @@ class GravatarOverlayView : UIView
         let ellipseRect = bounds.insetBy(dx: delta, dy: delta)
 
         // Setup
-        CGContextSaveGState(context)
-        CGContextSetLineWidth(context, borderWidth)
-        CGContextSetAllowsAntialiasing(context, true)
-        CGContextSetShouldAntialias(context, true)
+        CGContextSaveGState(context!)
```
Comment (koke) [inline WordPress/Classes/ViewRelated/Media/GravatarOverlayView.swift]: Can we just have `let context = UIGraphicsGetCurrentContext()!` above instead of adding `!`s all over the place?

File: WordPress/Classes/ViewRelated/NUX/SigninHelpers.swift (hunk 1 above)
Comment (koke) [inline WordPress/Classes/ViewRelated/NUX/SigninHelpers.swift]: Instead of `.characters.count == 0` we can use `.isEmpty`

File: WordPress/Classes/ViewRelated/Post/AztecPostViewController.swift [hunk 7]
```diff
@@ -11,7 +11,10 @@ class AztecPostViewController: UIViewController {
     static let margin = CGFloat(20)
 
     private(set) lazy var richTextView: Aztec.TextView = {
-        let tv = Aztec.TextView(defaultFont: WPFontManager.merriweatherRegularFontOfSize(16))
+        let defaultFont = WPFontManager.merriweatherRegularFontOfSize(16)
+        // TODO: Add a proper defaultMissingImage
```
Comment (koke) [inline WordPress/Classes/ViewRelated/Post/AztecPostViewController.swift]: Was this a TODO for the PR? Can we create a new issue instead?

File: WordPress/Classes/ViewRelated/Post/AztecPostViewController.swift [hunk 8]
```diff
@@ -404,7 +407,114 @@ extension AztecPostViewController : Aztec.FormatBarDelegate
 
 
     func toggleLink() {
-        richTextView.toggleLink(range: richTextView.selectedRange, params: [String : AnyObject]())
+        var linkTitle = ""
```
Comment (koke) [inline WordPress/Classes/ViewRelated/Post/AztecPostViewController.swift]: The following looks like a lot of new code for a swift 2.3 migration, can we split that into a separate PR? I'm not sure why it's here.

File: WordPress/Classes/ViewRelated/Reader/ReaderSearchViewController.swift [hunk 9]
```diff
@@ -241,7 +241,7 @@ import Gridicons
 
         // Match the width of the search bar.
         view.addConstraints(NSLayoutConstraint.constraintsWithVisualFormat("[autoView(==searchBar)]",
-            options: .AlignAllBaseline,
+            options: .AlignAllLastBaseline,
```
Comment (koke) [inline WordPress/Classes/ViewRelated/Reader/ReaderSearchViewController.swift]: This looks unrelated to Swift 2.3

**Comment by sendhil** [comment]
@jleandroperez - this was branched off of release/6.5 and was supposed to be merged into that 😁

**Comment by sendhil** [comment]
@koke those changes to the atzec files were just to get it to compile because a bunch of changes took place on the branch I fixed that broke it on wpios.

File: WordPress/Classes/ViewRelated/Reader/ReaderSearchViewController.swift (hunk 9 above)
Comment (sendhil) [inline WordPress/Classes/ViewRelated/Reader/ReaderSearchViewController.swift]: Threw an error until I changed it.

File: WordPress/Classes/ViewRelated/Post/AztecPostViewController.swift (hunk 8 above)
Comment (sendhil) [inline WordPress/Classes/ViewRelated/Post/AztecPostViewController.swift]: See my comment below about Aztec

File: WordPress/Classes/ViewRelated/Post/AztecPostViewController.swift (hunk 7 above)
Comment (sendhil) [inline WordPress/Classes/ViewRelated/Post/AztecPostViewController.swift]: It was more just for me but I believe @diegoreymendez is going to be cleaning up the Aztec issues here.

**Comment by sendhil** [comment]
Heads up - I changed the base.

File: WordPress/Classes/ViewRelated/Reader/ReaderPostCardCell.swift (hunk 3 above)
Comment (sendhil) [inline WordPress/Classes/ViewRelated/Reader/ReaderPostCardCell.swift]: Fixed in 6c5297b.

File: WordPress/Classes/ViewRelated/Reader/ReaderDetailViewController.swift (hunk 2 above)
Comment (sendhil) [inline WordPress/Classes/ViewRelated/Reader/ReaderDetailViewController.swift]: Fixed in 6c5297b

File: WordPress/Classes/ViewRelated/NUX/SigninHelpers.swift (hunk 1 above)
Comment (sendhil) [inline WordPress/Classes/ViewRelated/NUX/SigninHelpers.swift]: Fixed in 50d5f3b

**Comment by sendhil** [comment]
I changed base back - something happened with this branch since I pushed it up which is making it not want to merge into release/6.5. @jleandroperez did you merge develop into this branch?

**Comment by sendhil** [comment]
Ahhh shoot, I see the issue - https://github.com/wordpress-mobile/WordPress-iOS/pull/5977/commits/dbf40c91b9f5a2b088874c53e56647bcd0e9ebab was merged into here by mistake.

**Comment by sendhil** [comment]
Hmm, even with the revert commit it still is hosed when trying to merge into release. We might need to reconstruct this.

**Comment by sendhil** [comment]
Going to close this down in favor of https://github.com/wordpress-mobile/WordPress-iOS/pull/5983

**Comment by jleandroperez** [comment]
Wooow, thanks @sendhil!! lots and lots of activity!!!
